Add default block device support (SD, SPIF and FLASHIAP) documentation

yossi2le opened the pull request:
This PR is intended to update the documentation of block devices and filesystems after the merge of PR https://github.com/ARMmbed/mbed-os/pull/7774 into mbed-os.

The documentation contains 
   1. Change of paths.
   2. Update new implementation of get default instance for BlockDevice and FileSystem.
   3. Importing into mbed-os and modify the documentation of external SD, SPIF, DATAFLASH and FLASHIAP drivers 

Please refer to PR #7774 in the link above for all details regarding this change.

AnotherButler commented on docs/reference/api/storage/SDBlockDevice.md:
```diff
@@ -0,0 +1,115 @@
+- The file system code (hosted in `mbed-os`) is composed of 2 parts: the FAT file system implementation code and the file system classes that present a consistent API to the retarget module for different (third-party) file system implementations.
+- The Block API. The SDCard block device is a persistent storage block device.
+- The SPI module provides the Mbed OS SPI API.
+
```
Please add a section here linking to our Doxygen.

ashok-rao commented on docs/reference/api/storage/BlockDevice.md:
```diff
@@ -30,6 +30,66 @@ The state of an erased block is **undefined**. The data stored on the block isn'
+Please note that while a default block device exists an application is not enforced to use it and can create its own one.
+
+### Overriding default block device implementation 
+The get default instance is implemented as MBED_WEAK at features/storage/system_storage/SystemStorage.cpp. That means that it can be overridden by implementing the function without MBED_WEAK and change the default block device for a given application.
```
Maybe a link to -> [this](https://github.com/ARMmbed/mbed-os/blob/40058871de290edc758a21ae6d8f2ec1d1b3533d/platform/mbed_toolchain.h#L120) will be useful ?? it's not immediately clear to me what the macro means..

yossi2le commented:
Hi @AnotherButler.  I will update the PR today with changes addressing all comments.
@ashok-rao I am moving the configuration and overriding to storage configuration guide. I will add the link you ask over there. 
Thank you all, stay tuned.

yossi2le commented on docs/reference/api/storage/FlashIAPBlockDevice.md:
```diff
@@ -0,0 +1,17 @@
+
+## FlashIAPBlockDevice class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](<Should be added after doxygen run>)
```
no, but I will add a code snippet to help people understand how to use it.

yossi2le commented on docs/reference/api/storage/FlashIAPBlockDevice.md:
```diff
@@ -0,0 +1,17 @@
+
+The flash IAP block device is a block device driver built on top of the FlashIAP API.
+
+## Warning
```
This warning should be removed.

AnotherButler commented on docs/reference/configuration/Storage.md:
```diff
@@ -94,3 +96,97 @@ Name: nvstore.max_keys
+
+#### Configuring components
+
+Adding "components": ["???"] in targets.json:
```
@theotherjimmy As it's the weekend in Israel, could you please tell me what should go here?

theotherjimmy replied:
I think that's going to be one of "SD", "SPIF" or "FLASHIAP"

theotherjimmy replied:
This should just say "for example, the following entry in `targets.json` enables the SD component". The following example should be trimmed to only identifying info and the important information.

theotherjimmy replied:
```
 "K64F": {
        "components": ["SD"],
        "core": "Cortex-M4F",
        "supported_toolchains": ["ARM", "GCC_ARM", "IAR"],
        "inherits": ["Target"],
        "features": ["STORAGE"],
        "release_versions": ["2", "5"],
        ...
    },
```

Is probably enough to get the point across.

AnotherButler commented on docs/reference/configuration/Storage.md:
````diff
@@ -94,3 +96,97 @@ Name: nvstore.max_keys
+    },
+```
+
+Adding "target.components_add": ["???"] in application config file:
````
@theotherjimmy As it's the weekend in Israel, could you please tell me what should go here?

theotherjimmy replied:
Same as above.

AnotherButler replied:
So could I change it to

Adding "target.components_add": ["<SD>"] in application config file:

For clarity?

theotherjimmy replied:
No. That particular line is somewhat nonsense, somewhat leads to unexpected behavior.

theotherjimmy replied:
Woof. Alright.

"The following `mbed_app.json` snippet enables the Spi flash component when compiling for the MTB_ADV_WISE_1570 target.

```
       "MTB_ADV_WISE_1570": {
            "target.components_add": ["SPIF"]
       }
```

(removed unrelated garbage from the example)

AnotherButler replied:
@theotherjimmy Let's keep the comments professional and kind. I know @yossi2le and I both appreciate your help streamlining the example, but it doesn't mean the redundant lines are "garbage".

theotherjimmy replied:
As in "things to be thrown out"? I could have picked a more constructive word for sure. That word may have been in poor taste. Edited.

Personally I think unrelated lines take away from an example as they have the potential to distract or misdirect the reader and I find that having them is a determent to the example.

I meat no ill will or disparaging remarks to anyone or their writing skills/style.

yossi2le replied:
@theotherjimmy No hard feelings;)

AnotherButler commented on docs/reference/api/storage/DataFlashBlockDevice.md:
```diff
@@ -0,0 +1,49 @@
+
+### DataFlashBlockDevice class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](<Should be added after doxygen run>)
```
Note to self: Add link after code merges.

AnotherButler commented on docs/reference/api/storage/FlashIAPBlockDevice.md:
```diff
@@ -0,0 +1,46 @@
+
+## FlashIAPBlockDevice class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](<Should be added after doxygen run>)
```
Note to self: Add link after code merges.

AnotherButler commented on docs/reference/api/storage/SDBlockDevice.md:
```diff
@@ -0,0 +1,119 @@
+
+### SDBlockDevice class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](<Should be added after doxygen run>)
```
Note to self: Add link after code merges.

AnotherButler commented on docs/reference/api/storage/SPIFBlockDevice.md:
```diff
@@ -0,0 +1,44 @@
+
+### SPIFBlockDevice class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](<Should be added after doxygen run>)
```
Note to self: Add link after code merges.

yossi2le commented:
@AnotherButler  Thanks for the changes.

AnotherButler commented:
@yossi2le I can't find the code in the Doxygen: https://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/annotated.html

 Could you please help me?

yossi2le commented:
>@yossi2le I can't find the code in the Doxygen: https://os-doc->builder.test.mbed.com/docs/development/mbed-os-api-doxy/annotated.html
>
>Could you please help me?

I wish I could. I am puzzled, are we looking at the correct build.
its written at the end of the BlockDevice documentation:
```
The documentation for this class was generated from the following file:
mbed-os/features/filesystem/bd/BlockDevice.h
```
However, this file has been moved to mbed-os/features/storage/blockdevice/BlockDevice.h

I have a feeling this Doxygen output is before #7774

AnotherButler commented:
@kegilbert Is this something you can help with?

AnotherButler commented:
@yossi2le @kegilbert I'm going to comment out the broken links until we can fix them. That way, I can merge this, so OOB can start reviewing the descriptions and examples. Please do *not* stop working to figure out why these links don't work just because this PR is closing.

yossi2le commented:
@AnotherButler 
Like I wrote before I am guessing the links are broken because the Doxygen generated not from the most updated master.

AnotherButler commented:
@yossi2le @kegilbert I've updated the Doxygen, and these classes still aren't appearing: https://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/annotated.html

Should they be under Data Structures, where I'm looking? Or are they somewhere else?
